app/src/main/java/com/example/bigowlapp/repository/NotificationRepository.java

```diff
@@ -14,48 +11,32 @@
-                                                                                            Class<? extends SupervisionRequest> tClass) {
-        MutableLiveData<List<SupervisionRequest>> listOfTData = new MutableLiveData<>();
-        collectionReference.whereEqualTo(attribute, attrValue)
-                .whereEqualTo(TYPE, Notification.Type.SUPERVISION_REQUEST)
-                .get()
-                .addOnCompleteListener(task -> {
-                    if (task.isSuccessful()) {
-                        QuerySnapshot tDocs = task.getResult();
-                        if (tDocs != null && !tDocs.isEmpty()) {
-                            List<SupervisionRequest> listOfT = new ArrayList<>();
-                            for (QueryDocumentSnapshot doc : task.getResult()) {
-                                SupervisionRequest t = doc.toObject(tClass);
-                                listOfT.add(t);
-                            }
-                            listOfTData.setValue(listOfT);
-                        } else {
-                            listOfTData.setValue(null);
-                        }
-                    } else {
-                        Log.e(tClass.getSimpleName(), "Error getting documents: " +
-                                task.getException());
-                    }
-                });
-        return listOfTData;
+    public NotificationRepository(String userUid) {
+        super("users");
+        collectionReference = collectionReference.document(userUid).collection("notifications");
     }
 
     @Override
-    protected List<Notification> extractListOfDataToModel(QuerySnapshot results, Class<? extends Notification> tClass) {
-        List<Notification> notificationsFromDb = new ArrayList<>();
+    protected <X extends Notification> List<X> extractListOfDataToModel(QuerySnapshot results, Class<X> tClass) {
+        List<X> notificationsFromDb = new ArrayList<>();
         for (QueryDocumentSnapshot doc : results) {
-            Notification.Type type = doc.toObject(Notification.class).getType();
-            Notification t = doc.toObject(type.typeClass);
-            notificationsFromDb.add(t);
+            Notification notification = getNotificationFromDocument(doc, tClass);
+            if (notification.isValid()) {
+                notificationsFromDb.add((X) notification);
+            }
         }
         return notificationsFromDb;
     }
+
+    public static <X extends Notification> Notification getNotificationFromDocument(DocumentSnapshot doc, Class<X> xClass) {
+        Notification notification = Notification.getNotificationSafe(doc.toObject(Notification.class));
+        Notification.Type type = notification.getType();
+
+        if (xClass == Notification.class || xClass == type.typeClass) {
+            return doc.toObject(type.typeClass);
+        }
+
+        return new NullNotification();
+    }
```

Comment:
Excellent way to refactor the utils file authFailureNotificationListener to this. It is much simpler to put the logic here and database call here. Another extraction refactoring.